src/core/internal/dassert.d

```diff
@@ -377,7 +377,7 @@ private string miniFormat(V)(const scope ref V v)
 import core.atomic : atomicLoad;
 
 // Inverts a comparison token for use in _d_assert_fail
-private string invertCompToken(string comp) pure nothrow @nogc @safe
+private string invertCompToken(scope string comp) pure nothrow @nogc @safe
```

Comment:
I'm not sure if this should compile. Later in the function, we have `[comp]`. I.e., `comp` is copied to the heap. Usually, that's not allowed for `scope` variables.

Reply:
I think  it works here because the array is assigned to a `scope` parameter and hence is promoted to the stack.

Reply:
The `[comp]` is passed to parameter `const scope string[] valB` of `combine`, so the literal is allocated on the stack

Reply:
> I assume it works here because the array is assigned to a `scope` parameter and hence promoted to the stack.

That might be DMD's reasoning. But it's unsound. It actually doesn't matter if the array is on the heap or on the stack. A `scope` is being lost.

`comp.ptr` is a `scope` pointer. `valB.ptr` is also a `scope` pointer. `valB[0].ptr` is not a `scope` pointer, because `scope` is not transitive. So `valB = [comp];` cannot be allowed. It's assigning a `scope` pointer to a non-`scope` pointer.

Reduced code demonstrating memory corruption:

```d
string invertCompToken(scope string comp) @safe
{
    return combine([comp]); /* Should not compile. */
}
string combine(const scope string[] valB) @safe
{
    return valB[0];
}
string f() @safe
{
    immutable char[3] eq = "foo";
    return invertCompToken(eq);
}
void main() @safe
{
    import std.stdio;
    string s = f();
    writeln(s); /* prints garbage */
}
```

Reply:
Fair enough, that's just another oddity of `scope` not being transitive.
But there's no corruption here because `combine` doesn't escape the string. So it should be fine to wrap the call in a `@trusted` lambda to avoid future problems.

Reply:
I put the operator in the middle argument of `combine`, side-stepping the issue.

Reply:
Btw nice catch @aG0aep6G! I think the relevant bugzilla issue is this:
https://issues.dlang.org/show_bug.cgi?id=20505